e2etests/node_scan_rhcos_test.go

```diff
@@ -0,0 +1,186 @@
+		OsImage:          "Red Hat Enterprise Linux CoreOS 45.82.202008101249-0 (Ootpa)",
+		KernelVersion:    "0.0.1",
+		KubeletVersion:   "0.0.1",
+		KubeproxyVersion: "0.0.1",
+		Runtime: &v1.GetNodeVulnerabilitiesRequest_ContainerRuntime{
+			Name:    "docker",
+			Version: "0.0.1",
+		},
+		Notes: notes,
+		Components: &v1.Components{
+			Namespace: "rhcos:4.11",
+			RhelComponents: []*v1.RHELComponent{
+				{
+					Id:        int64(1),
+					Name:      "libksba",
+					Namespace: "rhel:8",
+					Version:   "1.3.5-7.el8",
+					Arch:      "x86_64",
+					Module:    "", // must be empty, otherwise scanner does not return any vulns
+					Cpes:      cpes,
+					AddedBy:   "",
+				},
+				{
+					Id:        int64(2),
+					Name:      "tar",
+					Namespace: "rhel:8",
+					Version:   "1.27.1.el8",
+					Arch:      "x86_64",
+					Module:    "",
+					Cpes:      cpes,
+					AddedBy:   "",
+				},
+				{
+					Id:        int64(3),
+					Name:      "grep",
+					Namespace: "rhel:8",
+					Version:   "3.1-6.el8",
+					Arch:      "x86_64",
+					Module:    "",
+					Cpes:      cpes,
+					AddedBy:   "",
+				},
+			},
+			LanguageComponents: nil,
+		},
+	}
+}
+
+func TestGRPCGetRHCOSNodeVulnerabilities(t *testing.T) {
+	conn := connectToScanner(t)
+	client := v1.NewNodeScanServiceClient(conn)
+
+	cases := []struct {
+		name             string
+		request          *v1.GetNodeVulnerabilitiesRequest
+		responseContains *v1.GetNodeVulnerabilitiesResponse
+	}{
+		{
+			name:    "Certified scan",
+			request: buildRequestCase([]v1.Note{}),
+			responseContains: &v1.GetNodeVulnerabilitiesResponse{
+				Features: []*v1.Feature{
+					{
+						Name:            "libksba",
+						Version:         "1.3.5-7.el8.x86_64",
+						Vulnerabilities: []*v1.Vulnerability{vulnLibksba},
+					},
+					{
+						Name:            "tar",
+						Version:         "1.27.1.el8.x86_64",
+						Vulnerabilities: []*v1.Vulnerability{vulnTar},
+					},
+					{
+						Name:            "grep",
+						Version:         "3.1-6.el8.x86_64",
+						Vulnerabilities: []*v1.Vulnerability{},
+					},
+				},
+			},
+		},
+		{
+			name:    "Uncertified scan is unsupported for RHCOS and returns no features",
+			request: buildRequestCase([]v1.Note{v1.Note_CERTIFIED_RHEL_SCAN_UNAVAILABLE}),
+			responseContains: &v1.GetNodeVulnerabilitiesResponse{
+				Features: []*v1.Feature{},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			c := c
+			resp, err := client.GetNodeVulnerabilities(context.Background(), c.request)
+			require.NoError(t, err)
+			assert.Len(t, resp.GetFeatures(), len(c.responseContains.GetFeatures()))
+			for _, expectedFeat := range c.responseContains.GetFeatures() {
+				var found bool
+				for _, gotFeat := range resp.GetFeatures() {
+					if expectedFeat.GetName() == gotFeat.GetName() && expectedFeat.GetVersion() == gotFeat.GetVersion() {
+						found = true
+						assert.NotNil(t, gotFeat)
+						if len(expectedFeat.GetVulnerabilities()) > 0 {
```

Comment:
One other thing to make this check more reliable would be to ignore any vulnerabilities for `gotFeat` that are newer than a hardcoded date of, say, 2022-12-07 that you can put in this test, if there is such a date attribute.

This way, you can even check equality of `gotFeat.GetVulnerabilities()` and `expectedFeat.GetVulnerabilities()` (ignoring order of elements).

Does something like creation date exist for vulnerabilities?
Judging by the name, `LastModifiedDateTime` seems more like an update date, and it may change for already known vulns causing the test to start failing when that happens.

Reply:
I can check the creation date, but the reason I am checking only for one vulnerability out of many is to save some space. I am not sure how many vulnerabilities exist for that version, I picked one as an example and I am asserting on that one.

All in all we are not testing here if Scanner returns the correct set of vulnerabilities, but if the API under test returns something that makes sense.

Reply:
I thought that you check for all vulns and the code did not tell otherwise. I think that deserves a comment somewhere around the place you define expected vulns that this is a spot-checking of vulns just to make sure some results come back.